hwcomposer/modes/waydroid_mode_impl.h

```diff
@@ -0,0 +1,139 @@
+/*
+* Copyright © 2025 Waydroid Project.
+*
+* Permission is hereby granted, free of charge, to any person obtaining
+* a copy of this software and associated documentation files (the
+* "Software"), to deal in the Software without restriction, including
+* without limitation the rights to use, copy, modify, merge, publish,
+* distribute, sublicense, and/or sell copies of the Software, and to
+* permit persons to whom the Software is furnished to do so, subject to
+* the following conditions:
+*
+* The above copyright notice and this permission notice (including the
+* next paragraph) shall be included in all copies or substantial
+* portions of the Software.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+* NONINFRINGEMENT.  IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+* BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+* ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+* CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+* SOFTWARE.
+ */
+
+#pragma once
+
+#include "waydroid_mode.h"
+
+template<class T>
+class compositing_window_mode : public virtual waydroid_mode {
+    skipped_layers_helper skipped_layers;
+
+    T *derived() {
+        return static_cast<T *>(this);
+    }
+
+  public:
+    int setup(waydroid_hwc_composer_device_1 *, hwc_display_contents_1_t *contents) override {
+        for (size_t i = 0; i < contents->numHwLayers; ++i) {
+            skipped_layers.setup_for_each_layer(i, contents->hwLayers[i]);
+        }
+        return 0;
+    }
+    int handle_layer(waydroid_hwc_composer_device_1 *pdev, hwc_layer_1 *hwc_layer, size_t i) override {
+        if (i == skipped_layers.first_skipped()) {
+            // Draw the content composited by SurfaceFlinger here
+            if (hwc_layer->acquireFenceFd != -1) {
+                close(hwc_layer->acquireFenceFd);
+            }
+
+            hwc_layer = skipped_layers.framebuffer_target_layer();
+            assert(hwc_layer);
+            i = skipped_layers.framebuffer_target_index();
+        } else if (skipped_layers.first_skipped() < i && i <= skipped_layers.last_skipped()) {
+            if (hwc_layer->acquireFenceFd != -1) {
+                close(hwc_layer->acquireFenceFd);
+            }
+            return 0;
+        } else if (hwc_layer == skipped_layers.framebuffer_target_layer() && skipped_layers.has_skipped_layers()) {
```

Comment:
```suggestion
        } else if (hwc_layer == skipped_layers.framebuffer_target_layer()) {
```

In compositing mode we never want to render the framebuffer_target_layer, unless we have skipped layers that were composited into it, otherwise we're potentially overlaying stale buffers. Indeed with full-ui or single-app in subsurface compositing mode, I get a frozen image after the bootanimation (you might need to use shm buffers to reproduce)